File: apps/windows/TeslaSync.App/feature-views/IncidentForm/IncidentForm.Model.cs
using System.Globalization;
using System.Text.Json.Serialization;
using TeslaSync.App.Core.Data.State;
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.FeatureViews;

/// <summary>
/// Incident severity — the native mirror of the web <c>IncidentSeverity</c> union
/// (<c>'minor' | 'major' | 'critical'</c>, web/src/api/hooks/useIncidents.ts). The wire form is the
/// lower-case token the Go API validates (<c>ValidateIncidentSeverity</c>).
/// </summary>
public enum IncidentSeverity
{
    /// <summary>A low-impact incident (web <c>minor</c>) — the form default.</summary>
    Minor,

    /// <summary>A partial-degradation incident (web <c>major</c>).</summary>
    Major,

    /// <summary>A full-outage incident (web <c>critical</c>).</summary>
    Critical,
}

/// <summary>
/// Incident lifecycle status — the native mirror of the web <c>IncidentStatus</c> union
/// (<c>'investigating' | 'identified' | 'monitoring' | 'resolved'</c>). The wire form is the lower-case
/// token the Go API validates (<c>ValidateIncidentStatus</c>).
/// </summary>
public enum IncidentStatus
{
    /// <summary>Cause not yet known (web <c>investigating</c>) — the form default.</summary>
    Investigating,

    /// <summary>Cause identified, fix pending (web <c>identified</c>).</summary>
    Identified,

    /// <summary>Fix applied, watching for recovery (web <c>monitoring</c>).</summary>
    Monitoring,

    /// <summary>Incident closed (web <c>resolved</c>).</summary>
    Resolved,
}

/// <summary>Wire mapping for <see cref="IncidentSeverity"/> — UI-free so it is asserted headlessly.</summary>
public static class IncidentSeverities
{
    /// <summary>The lower-case API token for <paramref name="severity"/>.</summary>
    public static string ToWire(IncidentSeverity severity) => severity switch
    {
        IncidentSeverity.Minor => "minor",
        IncidentSeverity.Major => "major",
        IncidentSeverity.Critical => "critical",
        _ => "minor",
    };

    /// <summary>Parse an API token back to a <see cref="IncidentSeverity"/>; false for an unknown token.</summary>
    public static bool TryFromWire(string? wire, out IncidentSeverity severity)
    {
        switch (wire)
        {
            case "minor":
                severity = IncidentSeverity.Minor;
                return true;
            case "major":
                severity = IncidentSeverity.Major;
                return true;
            case "critical":
                severity = IncidentSeverity.Critical;
                return true;
            default:
                severity = IncidentSeverity.Minor;
                return false;
        }
    }
}

/// <summary>Wire mapping for <see cref="IncidentStatus"/> — UI-free so it is asserted headlessly.</summary>
public static class IncidentStatuses
{
    /// <summary>The lower-case API token for <paramref name="status"/>.</summary>
    public static string ToWire(IncidentStatus status) => status switch
    {
        IncidentStatus.Investigating => "investigating",
        IncidentStatus.Identified => "identified",
        IncidentStatus.Monitoring => "monitoring",
        IncidentStatus.Resolved => "resolved",
        _ => "investigating",
    };

    /// <summary>Parse an API token back to a <see cref="IncidentStatus"/>; false for an unknown token.</summary>
    public static bool TryFromWire(string? wire, out IncidentStatus status)
    {
        switch (wire)
        {
            case "investigating":
                status = IncidentStatus.Investigating;
                return true;
            case "identified":
                status = IncidentStatus.Identified;
                return true;
            case "monitoring":
                status = IncidentStatus.Monitoring;
                return true;
            case "resolved":
                status = IncidentStatus.Resolved;
                return true;
            default:
                status = IncidentStatus.Investigating;
                return false;
        }
    }
}

/// <summary>
/// The <c>POST /api/v1/status/incidents</c> request body — the native mirror of the web
/// <c>CreateIncidentPayload</c> the form fills (web/src/features/system/components/status/IncidentForm.tsx
/// → <c>useCreateIncident</c>). Every property carries an explicit snake_case <see cref="JsonPropertyNameAttribute"/>
/// so the wire shape matches the Go <c>IncidentCreatePayload</c> regardless of the shared serializer's naming
/// policy. <see cref="InitialMessage"/> is omitted when null (web <c>initial_message: … || undefined</c>);
/// <see cref="AffectedComponents"/> is always sent, even when empty.
/// </summary>
public sealed record IncidentCreateRequest(
    [property: JsonPropertyName("title")] string Title,
    [property: JsonPropertyName("severity")] string Severity,
    [property: JsonPropertyName("status")] string Status,
    [property: JsonPropertyName("affected_components")] IReadOnlyList<string> AffectedComponents,
    [property: JsonPropertyName("initial_message")]
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    string? InitialMessage);

/// <summary>One severity choice for the severity dropdown (value + localized label).</summary>
public sealed record IncidentSeverityOption(IncidentSeverity Value, string Label);

/// <summary>One status choice for the status dropdown (value + localized label).</summary>
public sealed record IncidentStatusOption(IncidentStatus Value, string Label);

/// <summary>
/// A localized transient message for the toast surface — the native analogue of the web <c>useToast</c>
/// <c>toast.success</c> / <c>toast.error</c> calls. <see cref="IsError"/> selects the error vs. success
/// presentation.
/// </summary>
public sealed record IncidentFormToast(string Message, bool IsError);

/// <summary>
/// The outcome of a single create attempt — the native analogue of the web <c>useCreateIncident</c> mutation
/// resolving. On success it carries no payload (the form just closes); on an HTTP fault it carries a classified
/// <see cref="Error"/> rather than throwing (web parity: the mutation resolves to a toast, never an unhandled
/// rejection).
/// </summary>
public sealed record IncidentFormSubmitOutcome(bool Success, RepositoryError? Error)
{
    /// <summary>A successful create.</summary>
    public static IncidentFormSubmitOutcome Ok() => new(true, null);

    /// <summary>A classified failure.</summary>
    public static IncidentFormSubmitOutcome Fail(RepositoryError error) => new(false, error);
}

/// <summary>
/// Canonical metadata, validation bounds, Segoe Fluent glyphs and i18n keys for the <c>IncidentForm</c>
/// surface — the native mirror of <c>web/src/features/system/components/status/IncidentForm.tsx</c>. The web
/// component ships literal copy; every literal is keyed here (with that literal as the English fallback) so the
/// native view and view-model stay free of inline strings and resolve through the i18n facade. UI-free so every
/// key + bound is asserted in tests.
/// </summary>
public static class IncidentFormRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event (P1/S11 contract).</summary>
    public const string Slug = "IncidentForm";

    /// <summary>Minimum trimmed title length enforced client-side (web <c>t.length &lt; 3</c>) and by the API.</summary>
    public const int TitleMinLength = 3;

    /// <summary>Maximum title length (web <c>maxLength={200}</c> and the Go <c>IncidentTitleMaxLen</c>).</summary>
    public const int TitleMaxLength = 200;

    /// <summary>Maximum initial-message length (web <c>maxLength={4000}</c> and the Go <c>IncidentMessageMaxLen</c>).</summary>
    public const int MessageMaxLength = 4000;

    /// <summary>Segoe Fluent "ReportDocument" glyph standing in for the web modal's incident icon.</summary>
    public const string Glyph = "\uE7C3";

    /// <summary>Segoe Fluent "ErrorBadge" glyph for the inline / toast failure line (the web error toast).</summary>
    public const string ErrorGlyph = "\uEA39";

    /// <summary>Segoe Fluent "Completed" glyph for the success toast (the web success toast).</summary>
    public const string SuccessGlyph = "\uE930";

    /// <summary>The severity values in web render order (minor, major, critical).</summary>
    public static IReadOnlyList<IncidentSeverity> SeverityOrder { get; } =
    [
        IncidentSeverity.Minor,
        IncidentSeverity.Major,
        IncidentSeverity.Critical,
    ];

    /// <summary>The status values in web render order (investigating, identified, monitoring, resolved).</summary>
    public static IReadOnlyList<IncidentStatus> StatusOrder { get; } =
    [
        IncidentStatus.Investigating,
        IncidentStatus.Identified,
        IncidentStatus.Monitoring,
        IncidentStatus.Resolved,
    ];

    /// <summary>Modal title (web <c>title="Log an incident"</c>).</summary>
    public static string ModalTitle(ILocalizer localizer) =>
        Require(localizer).GetString("status.incidents.form.title", "Log an incident");

    /// <summary>Title field label (web <c>Title</c>).</summary>
    public static string TitleLabel(ILocalizer localizer) =>
        Require(localizer).GetString("status.incidents.form.titleLabel", "Title");

    /// <summary>Title field prompt (web <c>prompt</c>).</summary>
    public static string TitlePrompt(ILocalizer localizer) =>
        Require(localizer).GetString(
            "status.incidents.form.titlePrompt", "e.g. Wall connector restart at 14:00");

    /// <summary>Severity field label (web <c>Severity</c>).</summary>
    public static string SeverityLabel(ILocalizer localizer) =>
        Require(localizer).GetString("status.incidents.form.severityLabel", "Severity");

    /// <summary>Status field label (web <c>Status</c>).</summary>
    public static string StatusLabel(ILocalizer localizer) =>
        Require(localizer).GetString("status.incidents.form.statusLabel", "Status");

    /// <summary>Affected-components field label (web <c>Affected components</c>).</summary>
    public static string ComponentsLabel(ILocalizer localizer) =>
        Require(localizer).GetString("status.incidents.form.componentsLabel", "Affected components");

    /// <summary>Affected-components helper note (web <c>(comma-separated, optional)</c>).</summary>
    public static string ComponentsHint(ILocalizer localizer) =>
        Require(localizer).GetString("status.incidents.form.componentsHint", "(comma-separated, optional)");

    /// <summary>Affected-components prompt (web <c>prompt</c>).</summary>
    public static string ComponentsPrompt(ILocalizer localizer) =>
        Require(localizer).GetString("status.incidents.form.componentsPrompt", "e.g. tesla, telemetry");

    /// <summary>Initial-message field label (web <c>Initial timeline message</c>).</summary>
    public static string MessageLabel(ILocalizer localizer) =>
        Require(localizer).GetString("status.incidents.form.messageLabel", "Initial timeline message");

    /// <summary>Initial-message helper note (web <c>(optional)</c>).</summary>
    public static string MessageHint(ILocalizer localizer) =>
        Require(localizer).GetString("status.incidents.form.messageHint", "(optional)");

    /// <summary>Initial-message prompt (web <c>prompt</c>).</summary>
    public static string MessagePrompt(ILocalizer localizer) =>
        Require(localizer).GetString("status.incidents.form.messagePrompt", "What\u2019s the situation?");

    /// <summary>Cancel button label (web <c>Cancel</c>).</summary>
    public static string CancelLabel(ILocalizer localizer) =>
        Require(localizer).GetString("status.incidents.form.cancel", "Cancel");

    /// <summary>Submit button label, idle (web <c>Log incident</c>).</summary>
    public static string SubmitLabel(ILocalizer localizer) =>
        Require(localizer).GetString("status.incidents.form.submit", "Log incident");

    /// <summary>Submit button label, busy (web <c>Logging\u2026</c>).</summary>
    public static string SubmittingLabel(ILocalizer localizer) =>
        Require(localizer).GetString("status.incidents.form.submitting", "Logging\u2026");

    /// <summary>Client-side title-too-short validation message (web <c>toast.error</c> copy).</summary>
    public static string TitleTooShortMessage(ILocalizer localizer) =>
        Require(localizer).GetString(
            "status.incidents.form.titleTooShort", "Title must be at least 3 characters.");

    /// <summary>Success toast (web <c>toast.success('Incident logged.')</c>).</summary>
    public static string SuccessMessage(ILocalizer localizer) =>
        Require(localizer).GetString("status.incidents.form.success", "Incident logged.");

    /// <summary>Failure toast fallback (web <c>'Failed to log incident'</c>).</summary>
    public static string ErrorMessage(ILocalizer localizer) =>
        Require(localizer).GetString("status.incidents.form.error", "Failed to log incident");

    /// <summary>The localized label for a severity value (web option labels).</summary>
    public static string SeverityLabelFor(IncidentSeverity severity, ILocalizer localizer) => severity switch
    {
        IncidentSeverity.Minor => Require(localizer).GetString("status.incidents.severity.minor", "Minor"),
        IncidentSeverity.Major => Require(localizer).GetString("status.incidents.severity.major", "Major"),
        IncidentSeverity.Critical => Require(localizer).GetString("status.incidents.severity.critical", "Critical"),
        _ => Require(localizer).GetString("status.incidents.severity.minor", "Minor"),
    };

    /// <summary>The localized label for a status value (web option labels).</summary>
    public static string StatusLabelFor(IncidentStatus status, ILocalizer localizer) => status switch
    {
        IncidentStatus.Investigating =>
            Require(localizer).GetString("status.incidents.status.investigating", "Investigating"),
        IncidentStatus.Identified =>
            Require(localizer).GetString("status.incidents.status.identified", "Identified"),
        IncidentStatus.Monitoring =>
            Require(localizer).GetString("status.incidents.status.monitoring", "Monitoring"),
        IncidentStatus.Resolved =>
            Require(localizer).GetString("status.incidents.status.resolved", "Resolved"),
        _ => Require(localizer).GetString("status.incidents.status.investigating", "Investigating"),
    };

    private static ILocalizer Require(ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(localizer);
        return localizer;
    }
}

/// <summary>
/// Pure projections for the <c>IncidentForm</c> surface — the native analogue of the web component's option
/// lists, comma-list parsing, client-side title validation and payload assembly. Every user-visible string
/// flows through the i18n facade so the projection is unit-tested headlessly and the view-model never resolves
/// a literal.
/// </summary>
public static class IncidentFormProjection
{
    /// <summary>The severity dropdown options in web render order with localized labels.</summary>
    public static IReadOnlyList<IncidentSeverityOption> SeverityOptions(ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(localizer);
        var options = new List<IncidentSeverityOption>(IncidentFormRegistration.SeverityOrder.Count);
        foreach (var value in IncidentFormRegistration.SeverityOrder)
        {
            options.Add(new IncidentSeverityOption(value, IncidentFormRegistration.SeverityLabelFor(value, localizer)));
        }

        return options;
    }

    /// <summary>The status dropdown options in web render order with localized labels.</summary>
    public static IReadOnlyList<IncidentStatusOption> StatusOptions(ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(localizer);
        var options = new List<IncidentStatusOption>(IncidentFormRegistration.StatusOrder.Count);
        foreach (var value in IncidentFormRegistration.StatusOrder)
        {
            options.Add(new IncidentStatusOption(value, IncidentFormRegistration.StatusLabelFor(value, localizer)));
        }

        return options;
    }

    /// <summary>The trimmed title (web <c>title.trim()</c>).</summary>
    public static string NormalizeTitle(string? title) => (title ?? string.Empty).Trim();

    /// <summary>True once the trimmed title meets the client-side minimum (web <c>t.length &gt;= 3</c>).</summary>
    public static bool IsTitleValid(string? title) =>
        NormalizeTitle(title).Length >= IncidentFormRegistration.TitleMinLength;

    /// <summary>
    /// Split a comma-separated component list into trimmed, non-empty tokens — the exact web transform
    /// (<c>components.split(',').map(trim).filter(Boolean)</c>).
    /// </summary>
    public static IReadOnlyList<string> ParseComponents(string? components)
    {
        if (string.IsNullOrWhiteSpace(components))
        {
            return Array.Empty<string>();
        }

        var result = new List<string>();
        foreach (var raw in components.Split(','))
        {
            string trimmed = raw.Trim();
            if (trimmed.Length > 0)
            {
                result.Add(trimmed);
            }
        }

        return result;
    }

    /// <summary>
    /// Assemble the create request from the current field values — the native analogue of the object the web
    /// passes to <c>create.mutateAsync</c>. The title is trimmed, the initial message is trimmed-or-null
    /// (omitted on the wire when empty) and the components are parsed into the affected-components array.
    /// </summary>
    public static IncidentCreateRequest BuildRequest(
        string? title,
        IncidentSeverity severity,
        IncidentStatus status,
        string? components,
        string? message)
    {
        string trimmedMessage = (message ?? string.Empty).Trim();
        return new IncidentCreateRequest(
            NormalizeTitle(title),
            IncidentSeverities.ToWire(severity),
            IncidentStatuses.ToWire(status),
            ParseComponents(components),
            trimmedMessage.Length > 0 ? trimmedMessage : null);
    }
}

/// <summary>
/// PII-safe diagnostics for the <c>IncidentForm</c> surface (P1/S11 diagnostics contract). Records only
/// operational counters with the surface slug — never the incident title, components or message — so a
/// diagnostics line can never leak incident content. Thread-safe.
/// </summary>
public sealed class IncidentFormDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;
    private long _incidentsLogged;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    public IncidentFormDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Number of incidents successfully logged from this surface.</summary>
    public long IncidentsLogged => Interlocked.Read(ref _incidentsLogged);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=IncidentForm</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke(string.Create(
            CultureInfo.InvariantCulture, $"view.opened slug={IncidentFormRegistration.Slug}"));
    }

    /// <summary>Record that an incident was logged (the title / components are never logged).</summary>
    public void RecordIncidentLogged()
    {
        Interlocked.Increment(ref _incidentsLogged);
        _sink?.Invoke(string.Create(
            CultureInfo.InvariantCulture, $"status.incident.logged slug={IncidentFormRegistration.Slug}"));
    }
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/IncidentForm/IncidentFormModel.kt
// Pure, framework-free model + projection for the IncidentForm feature view — the native analogue of everything
// the web component derives before it returns JSX (web/src/features/system/components/status/IncidentForm.tsx). No
// Compose, no Android, no HTTP: every declaration here is exercised off-device by the :android:testReleaseUnitTest
// gate, so the composable stays a thin render layer over these pure functions.
//
// The web component is a manual incident-logging dialog: a required title (length 3–200, validated client-side AND
// server-side), a severity selector (minor/major/critical), a status selector
// (investigating/identified/monitoring/resolved), an optional comma-separated affected-components field, and an
// optional initial timeline message. On submit it POSTs to `useCreateIncident` (the shared IncidentsStore, P1/S8)
// and closes; the list query is invalidated by the store. This file owns the data derivations behind that form: the
// title-length validation (web `t.length < 3`), the create-payload assembly (web object literal — trim title, drop a
// blank message to `null`, split + trim + drop-empty the components), and the severity/status option vocabularies.
// Colors, glyphs and localized labels are resolved at the Compose boundary, never here.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/feature-views/IncidentForm — the P3 prompt's allowed-files path) cannot form a valid Kotlin package
// (a hyphen and a PascalCase segment are illegal in a package identifier), so the package intentionally diverges from
// the path — exactly as the sibling feature-view surfaces do. `MatchingDeclarationName` is suppressed for the
// co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.featureviews.incidentform

import io.teslasync.shared.core.diagnostics.Logger
import io.teslasync.shared.core.presentation.incidents.CreateIncidentInput
import io.teslasync.shared.core.presentation.incidents.Incident
import io.teslasync.shared.core.presentation.incidents.IncidentsStore

/** Diagnostics + registry identifiers for the surface (P1/S11). */
object IncidentFormRegistration {
    /** Stable surface id. */
    const val ID: String = "incident-form"

    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11). */
    const val SLUG: String = "IncidentForm"
}

/**
 * The incident severity union the web component offers (`minor | major | critical`). [wire] is the exact lowercase
 * token sent to `POST /status/incidents`; the human label is resolved at the Compose boundary (P1/S10).
 */
enum class IncidentSeverity(
    val wire: String,
) {
    Minor("minor"),
    Major("major"),
    Critical("critical"),
    ;

    companion object {
        /** Resolves a [wire] token back to its case (web select `onChange`); unknown tokens fall back to [Minor]. */
        fun fromWire(wire: String): IncidentSeverity = entries.firstOrNull { it.wire == wire } ?: Minor
    }
}

/**
 * The incident lifecycle-status union the web component offers
 * (`investigating | identified | monitoring | resolved`). [wire] is the exact lowercase token sent to the server;
 * the human label is resolved at the Compose boundary (P1/S10).
 */
enum class IncidentStatus(
    val wire: String,
) {
    Investigating("investigating"),
    Identified("identified"),
    Monitoring("monitoring"),
    Resolved("resolved"),
    ;

    companion object {
        /** Resolves a [wire] token back to its case (web select `onChange`); unknown tokens fall back to [Investigating]. */
        fun fromWire(wire: String): IncidentStatus = entries.firstOrNull { it.wire == wire } ?: Investigating
    }
}

/**
 * The editable form draft the dialog owns — the native mirror of the web component's `useState` fields. Defaults
 * match the web initial state (`severity='minor'`, `status='investigating'`, the rest empty), so a freshly opened
 * dialog is immediately submittable once a title is typed.
 */
data class IncidentDraft(
    val title: String = "",
    val severity: IncidentSeverity = IncidentSeverity.Minor,
    val status: IncidentStatus = IncidentStatus.Investigating,
    val components: String = "",
    val message: String = "",
)

/**
 * The transient toasts the surface raises (web `useToast`), localized + toned at the Compose boundary (P1/S10).
 * Carries no pre-localized sentence and no PII (ADR-016) — [SubmitFailed.detail] holds the server-supplied error
 * text the web shows verbatim (`err.message`), which the boundary falls back to a localized message for when blank.
 */
sealed interface IncidentFormToast {
    /** Web `toast.success('Incident logged.')`. */
    data object Logged : IncidentFormToast

    /** Web `toast.error('Title must be at least 3 characters.')`. */
    data object ValidationTitleTooShort : IncidentFormToast

    /** Web `toast.error(err.message ?? 'Failed to log incident')`. */
    data class SubmitFailed(
        val detail: String?,
    ) : IncidentFormToast
}

/**
 * The pure derivations the composable renders over — the native mirror of the web component's inline submit logic.
 * Stateless and side-effect-free, so it is fully covered by the off-device unit gate.
 */
object IncidentFormProjection {
    /** Minimum trimmed title length the client enforces before submit (web `t.length < 3`). */
    const val MIN_TITLE_LENGTH: Int = 3

    /** Maximum title length the input accepts (web `maxLength={200}`), mirrored by the server bound. */
    const val MAX_TITLE_LENGTH: Int = 200

    /** Maximum initial-message length the textarea accepts (web `maxLength={4000}`). */
    const val MAX_MESSAGE_LENGTH: Int = 4000

    /** Whether [title] satisfies the client-side minimum once trimmed (web submit guard). */
    fun isTitleValid(title: String): Boolean = title.trim().length >= MIN_TITLE_LENGTH

    /** Clamps a title edit to the accepted maximum (web `maxLength={200}`). */
    fun clampTitle(title: String): String = title.take(MAX_TITLE_LENGTH)

    /** Clamps an initial-message edit to the accepted maximum (web `maxLength={4000}`). */
    fun clampMessage(message: String): String = message.take(MAX_MESSAGE_LENGTH)

    /**
     * Splits the comma-separated affected-components field into trimmed, non-empty tokens — the web
     * `components.split(',').map((c) => c.trim()).filter(Boolean)`. An all-blank field yields an empty list, exactly
     * as the web sends `affected_components: []`.
     */
    fun parseComponents(raw: String): List<String> = raw.split(',').map { it.trim() }.filter { it.isNotEmpty() }

    /**
     * Assembles the `POST /status/incidents` body from [draft] — the web `create.mutateAsync({...})` object literal.
     * The title is trimmed; the message is trimmed and dropped to `null` when blank (web `message.trim() ||
     * undefined`); the components are split + trimmed + de-blanked; the description is never sent by the dialog.
     */
    fun buildCreateInput(draft: IncidentDraft): CreateIncidentInput =
        CreateIncidentInput(
            title = draft.title.trim(),
            severity = draft.severity.wire,
            status = draft.status.wire,
            affectedComponents = parseComponents(draft.components),
            initialMessage = draft.message.trim().ifBlank { null },
        )
}

/**
 * The narrow write seam the dialog binds to — the native analogue of the web `useCreateIncident` mutation. A
 * production binding routes to the shared [IncidentsStore] (see [bindIncidentFormSource]); tests pass a fake. Keeping
 * the seam this small means the dialog never sees the store, the cache, or HTTP.
 */
interface IncidentFormSource {
    /** Creates an incident, returning the non-throwing [Result] the store exposes (web `mutateAsync`). */
    suspend fun createIncident(input: CreateIncidentInput): Result<Incident>
}

/**
 * Binds the dialog's write seam to the shared **S8** [IncidentsStore] (web `useCreateIncident`). The store refreshes
 * every observed incidents feed on success, so a successful log updates the status-page list with no extra wiring.
 */
fun bindIncidentFormSource(store: IncidentsStore): IncidentFormSource =
    object : IncidentFormSource {
        override suspend fun createIncident(input: CreateIncidentInput): Result<Incident> = store.createIncident(input)
    }

/**
 * Emits the one PII-safe `view.opened` diagnostic with the surface [IncidentFormRegistration.SLUG] (P1/S11). Kept
 * free of Compose so it is unit-tested with a recording [Logger]; the composable calls it from its first-composition
 * effect.
 */
fun recordIncidentFormOpened(logger: Logger) {
    logger.info("view.opened", mapOf("surface" to IncidentFormRegistration.SLUG))
}
